helm: support encryption config in ceph-csi-cephfs chart

acolombier opened the pull request:
This PR adds encryption configuration to the `cephfs-csi` chart, similarly to the RBD one,

This fixes permission missing when using the `kubernetes` KMS

## Related issues ##

Fixes: #https://github.com/ceph/ceph-csi/issues/4470

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [?] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [x] Documentation has been updated, if necessary.
* [-] Unit tests have been added, if necessary.
* [-] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

acolombier committed:
helm: support encryption config in ceph-csi-cephfs chart

this chart currently lack the ability to properly configure encryption,
as well as granting sufficent permission to allow controllers to access
secret when needed.

Signed-off-by: Antoine C <hi@acolombier.dev>

acolombier commented:
Let me know if you would need me to update the `Pending release notes` for this PR

nixpanic commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -30,6 +30,19 @@ serviceAccounts:
 csiConfig: []
 
+# Configuration for the encryption KMS
+# Ref: https://github.com/ceph/ceph-csi/blob/devel/docs/deploy-cephfs.md#cephfs-volume-encryption
```
this line seems to be too long ([ci job](https://github.com/ceph/ceph-csi/actions/runs/8514964926/job/23336366170?pr=4531#step:3:1116))

acolombier replied:
Is it okay for me to break the link into two lines? Or perhaps I should add lint silencer for this line? What is the suggestion?

nixpanic replied:
Adding a lint silencer for that line is fine. Breaking the URL over multiple lines isn't that clean.

nixpanic replied:
This line directly above the URL should work:
```
# yamllint disable-line rule:line-length
```

nixpanic commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -300,6 +313,18 @@ storageClass:
+  # specifying a unique ID matching KMS ConfigMap. The ID is only used for
+  # correlation to configmap entry.
+  encryptionKMSID: ""
+  
```
there seems to be a trailing space here ([ci job](https://github.com/ceph/ceph-csi/actions/runs/8514964926/job/23336366170?pr=4531#step:3:1116))

nixpanic reviewed (commented):
minor changes needed to pass CI/yamllint

nixpanic commented:
`yamllint` passed, now MarkDown linting needs to be fixed as well...

![image](https://github.com/ceph/ceph-csi/assets/400257/7d33a7b3-29aa-4a72-8049-ea636c087df7)

acolombier commented:
Apologies for missing this previously @nixpanic . Is it expected that the pre-commit check doesn't run any of the checkers? Or is my local setup flawed?

nixpanic commented:
> Apologies for missing this previously @nixpanic . Is it expected that the pre-commit check doesn't run any of the checkers? Or is my local setup flawed?

I don't think pre-commit runs all the checks. I usually run `make containerized-test` to check if everything is OK.

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-clusterrole.yaml:
```diff
@@ -3,6 +3,7 @@ kind: ClusterRole
 apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: {{ include "ceph-csi-cephfs.nodeplugin.fullname" . }}
+  namespace: {{ .Release.Namespace }}
```
we dont need namespace for clusterRole

Madhu-1 commented on charts/ceph-csi-cephfs/templates/nodeplugin-role.yaml:
```diff
@@ -0,0 +1,21 @@
+{{- if .Values.rbac.create -}}
+{{- if and .Values.encryptionKMSConfig .Values.encryptionKMSConfig.secretNamespace  .Values.encryptionKMSConfig.secretName (eq .Values.encryptionKMSConfig.secretNamespace .Release.Namespace) -}}
+kind: Role
```
we dont need Role and Role Binding right? we need to have access to secrets in different namespaces as well. i don't see Role https://github.com/ceph/ceph-csi/tree/devel/deploy/cephfs/kubernetes, am i missing anything?

acolombier replied:
The idea was to downgrade the `ClusterRole` to just a `Role` in case the configured secret in `encryptionKMSConfig` was local to the namespace. This was done to help with least privileged and limit the access to secret as much as possible. Happy to remove if you think this is overkill. Arguably, users who want least privileged roles can create manage the RBAC themselves

nixpanic replied:
Ceph-CSI components which reads the Secrets are usually running in a different namespace than the namespace where Secrets for applications have their encryption key. Ceph-CSI needs to read the Secret from a different namespace than where it is running/mounting the PV.

acolombier replied:
In this case, it will create the `ClusterRole` automatically. The logic at line 2 is if `
encryptionKMSConfig.secretNamespace == Release.Namespace`, then create a `Role` instead of adding `secret:read` to the `ClusterRole`.

acolombier replied:
Happy to scratch that optimisation, and create RBAC myself! But just thought it would be a harmless optimisation to put in

nixpanic replied:
I think the optimization is nice to have, specially if users (like yourself) will benefit from it. Maybe document it, or explain it in a comment, and add it to the rbd chart too?

acolombier replied:
No problem, I will put some comments in the README. I'll try to see if I can add it the the rbd too, is it okay if I do so in another PR tho? I seem to recall than `encryptionKMSConfig` doesn't work exactly the same and so it might be best to split the changes if that's okay with you?

nixpanic replied:
Thanks! Yes, a 2nd PR is fine for the RBD enhancement.

acolombier replied:
I have slightly reworked the logic behind it to enforce determinism and added a some documentation in the `REAMDE.md`. I'd be nice if you could tell me how that looks, and if you like it, I can spin up a PR with the same feature for RBD.

acolombier replied:
@nixpanic it'd be great if you could let me know your though on this

Madhu-1 replied:
@acolombier @nixpanic is on PTO will be back 1st week of Jun, do you want me to block this PR until he is back to confirm on this?

acolombier replied:
Up to you! If it is ready to be merged, happy to do so too, and I'll wait and see if @nixpanic likes this approach for RBD and I can make the PR then

acolombier replied:
@nixpanic friendly ping in case you missed the notification :)

acolombier committed:
helm: add least privileges logic for secrets on ceph-csi-cephfs chart

this allows the encryption KMS config to be granted secret access with
a least privilges policy.

Signed-off-by: Antoine C <hi@acolombier.dev>

Madhu-1 reviewed (dismissed):
LGTM, Thanks

acolombier commented:
Not sure what is going on with `commitlint` - is that something that needs to be fixed on my end? Do you want me to merge devel?

Edit my commit to fix mdlint - sorry @Madhu-1

Madhu-1 commented:
@acolombier can you please rebase the PR on top of devel branch?

acolombier commented:
Done @Madhu-1

Madhu-1 commented:
@mergifyio rebase

nixpanic reviewed (dismissed):
Sorry for the delayed response. It's a little sad none of the other contributors had a look in such a long time.

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

acolombier commented:
Thanks for not letting the PR die @nixpanic ! 

Let me know how/if I can fix those tests!

iPraveenParihar commented:
/retest ci/centos/mini-e2e-helm/k8s-1.29

nixpanic commented:
> Let me know how/if I can fix those tests!

It seems to go wrong with the cephfs provisioner deployment, and maybe other parts.

https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.29/detail/mini-e2e-helm_k8s-1.29/514/pipeline/90

I suspect something sneaked in the helm charts that causes some problematic yaml generation.

Probably this:
```
Error: INSTALLATION FAILED: Unable to continue with install: could not get information about the resource ConfigMap "" in namespace "cephcsi-e2e-aa06ab4c126c": resource name may not be empty
```

acolombier commented:
Thanks for the tip @nixpanic - indeed, as I tried to unify the way encryption works with the RBD chart, I forgot some chart values from there. Should be ready for testing now!

Will squash once the test passes if that's okay with you.

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.29

iPraveenParihar commented:
```
Error: INSTALLATION FAILED: Unable to continue with install: ConfigMap "ceph-csi-encryption-kms-config" in namespace "cephcsi-e2e-d1df94a5c755" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "ceph-csi-rbd": current value is "ceph-csi-cephfs"
```

@acolombier, the install-helm.sh script fails to install RBD driver. Could you please include the change to delete the configMap `ceph-csi-encryption-kms-config` in install-helm.sh?

https://github.com/ceph/ceph-csi/blob/cea8bf8110531284d272582b9c536ffdb2ece82b/scripts/install-helm.sh#L187-L191

acolombier commented:
Done @iPraveenParihar !

iPraveenParihar commented:
/test ci/centos/mini-e2e-helm/k8s-1.29

iPraveenParihar commented:
the helm test passed 🎉

iPraveenParihar commented:
@acolombier, can you please squash your `fixup!` commits?

acolombier commented:
Done!

iPraveenParihar commented:
@Mergifyio queue

acolombier commented:
> Please make sure `acolombier` has logged in Mergify dashboard

I have logged in with Mergify now. Hopefully this is enough to merge @iPraveenParihar ?

iPraveenParihar commented:
@Mergifyio queue

iPraveenParihar commented:
@Rakshith-R, I think there is some issue with me running Mergifyio queue. Can you run it from your end?

Rakshith-R commented:
@mergifyio rebase

iPraveenParihar commented:
@Mergifyio rebase

iPraveenParihar commented:
@Mergifyio queue
